packages/lib/server/repository/bookingReference.ts

```diff
@@ -49,4 +57,106 @@ export class BookingReferenceRepository {
       }),
     });
   }
+
+  /**
+   * Whenever a new Booking is created, a BookingReference is also created and is connected with Credential record by 'CredentialId'.
+   * If for some reason a App is uninstalled and installed, the Credential record is deleted and BookingReference is orphaned by 'CredentialId' field being set to null.
+   * This function detects the orphaned BookingReference whenever a new Credential is created and reconnects it.
+   * So, this function has to be called whenever a new Credential record is created.
+   * The combination of 3 fields - 'userId or teamId' , 'credentialType' , 'credentialId==null' is used to detect orphaned bookingReference.
+   * For Calendar Apps, additional field - 'externalCalendarId' is used to detect orphaned bookingReferences.
+   * Hence, this function has to be called also whenever a new SelectedCalendar (linked with credential) record is added.
+   */
+  static async reconnectWithNewCredential(newCredentialId: number) {
+    try {
+      const newCredential = await CredentialRepository.findByIdWithSelectedCalendar({
+        id: newCredentialId,
+      });
+
+      if (!newCredential) {
+        throw new Error("Credential not found.");
+      }
+
+      //In case of credential created for a team, get member userIds
+      let teamMembersUserIds: number[] = [];
+      if (newCredential.teamId && !newCredential.userId) {
+        const members = await prisma.membership.findMany({
+          where: {
+            teamId: newCredential.teamId,
+          },
+          select: {
+            userId: true,
+          },
+        });
+        teamMembersUserIds = members.map((member) => member.userId);
+      }
+
+      //Detect bookingReferences to connect with new Credential.
+      const bookingReferences = await prisma.bookingReference.findMany({
+        where: {
+          type: newCredential.type,
+          booking: {
+            ...(newCredential.userId
+              ? { userId: newCredential.userId }
+              : { userId: { in: teamMembersUserIds } }),
+          },
+          credentialId: null,
+          ...(!newCredential.selectedCalendars || newCredential.selectedCalendars.length === 0
+            ? { externalCalendarId: null } // for non-calendar apps
+            : {
+                externalCalendarId: {
+                  in: newCredential.selectedCalendars
+                    .filter((selectedCalendar) => !!selectedCalendar.externalId)
+                    .map((selectedCalendar) => selectedCalendar.externalId as string),
+                },
+              }),
+        },
+        select: {
+          id: true,
+          bookingId: true,
+        },
+      });
```

Comment:
raw SQL Query generated -
```
SELECT
    "BookingReference"."id",
    "BookingReference"."type"
FROM
    "public"."BookingReference"
LEFT JOIN
    "public"."Booking" AS "j1" ON "j1"."id" = "BookingReference"."bookingId"
WHERE
    "BookingReference"."type" = $1
    AND "j1"."userId" = $2
    AND "j1"."id" IS NOT NULL
    AND "BookingReference"."credentialId" IS NULL
    AND "BookingReference"."externalCalendarId" IN ($3)
    AND "BookingReference"."deleted" IS NULL
OFFSET $4;
```
IMO: results in single left-join, not much complex.